File: 03_Sliding_Window/03_Hard/minimumWindowSubstring.py
'''
Given two strings s and t of lengths m and n respectively, return the minimum window 
substring of s such that every character in t (including duplicates) is included in the window. 
If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.
'''

from collections import defaultdict


def inside(chars, window):
    for key, value in chars.items():
        if key not in window or window[key] == 0:
            return False
        elif window[key] < value:
            return False

    return True


def minWindow(s, t):
    tCount = defaultdict(int)
    for char in t:
        tCount[char] += 1

    windowCount = defaultdict(int)
    res = ""
    l = 0

    for r in range(len(s)):
        windowCount[s[r]] += 1

        while inside(tCount, windowCount):
            windowCount[s[l]] -= 1
            l += 1

            #res = s[l-1:r+1]
            if res == "" or len(res) > len(s[l-1:r+1]):
                res = s[l-1:r+1]

    return res

# First, store a frequency count of all the characters in t. This will be used in the helper 
# function which checks if one dictionary is a subset of another. Next, create a window. Increase
# the right index until the window is valid, at which point, increase the left window until the window 
# is no longer valid. This will allow us to get all valid windows. Keep the shorted valid window. 

# Time | Space : O(n) | O(n)

File: 03_Sliding_Window/03_Hard/C++/minimumWindowSubstring.cpp
#include <string>
#include <unordered_map>

class Solution {
public:
    std::string minWindow(std::string s, std::string t) {
        
        auto inside = [](const std::unordered_map<char, int>& charCount, const std::unordered_map<char, int>& window) {
            for (const auto& [key, value] : charCount) {
                if (window.find(key) == window.end() || window.at(key) == 0 || window.at(key) < value) {
                    return false;
                }
            }
            return true;
        };

        std::unordered_map<char, int> tCount;
        for (char charT : t) {
            tCount[charT]++;
        }

        std::unordered_map<char, int> windowCount;
        std::string res = "";
        int l = 0;

        for (int r = 0; r < s.length(); r++) {
            windowCount[s[r]]++;

            while (inside(tCount, windowCount)) {
                windowCount[s[l]]--;
                l++;

                if (res == "" || res.length() > r - l + 2) {
                    res = s.substr(l - 1, r - l + 2);
                }
            }
        }

        return res;
    }
};
